mmgen/models/diffusions/base_diffusion.py

```diff
@@ -0,0 +1,903 @@
+        r"""Create a beta schedule that discretizes the given alpha_t_bar
+        function, which defines the cumulative product of `(1-\beta)` over time
+        from `t = [0, 1]`.
+
+        Args:
+            diffusion_timesteps (int): The number of betas to produce.
+            max_beta (float, optional): The maximum beta to use; use values
+                lower than 1 to prevent singularities. Defaults to 0.999.
+            s (float, optional): Small offset to prevent `\beta` from being too
+                small near `t = 0` Defaults to 0.008.
+
+        Returns:
+            np.ndarray: Betas used in diffusion process.
+
+        """
+
+        def f(t, T, s):
+            return np.cos((t / T + s) / (1 + s) * np.pi / 2)**2
+
+        betas = []
+        for t in range(diffusion_timesteps):
+            alpha_bar_t = f(t + 1, diffusion_timesteps, s)
+            alpha_bar_t_1 = f(t, diffusion_timesteps, s)
+            betas_t = 1 - alpha_bar_t / alpha_bar_t_1
+            betas.append(min(betas_t, max_beta))
+        return np.array(betas)
+
+    def q_sample(self, x_0, t, noise=None):
+        r"""Get diffusion result at timestep `t` by `q(x_t | x_0)`.
+
+        Args:
+            x_0 (torch.Tensor): Original image without diffusion.
+            t (torch.Tensor): Target diffusion timestep.
+            noise (torch.Tensor, optional): Noise used in reparameteration
+                trick. Default to None.
+
+        Returns:
+            torch.tensor: Diffused image `x_t`.
+        """
+        num_batches = x_0.shape[0]
+        tar_shape = x_0.shape
+        # noise = self._get_noise_batch(noise, num_batches, noise_size)
+        noise = self.get_noise(noise, num_batches=num_batches)
+        mean = var_to_tensor(self.sqrt_alphas_bar, t, tar_shape)
+        std = var_to_tensor(self.sqrt_one_minus_alphas_bar, t, tar_shape)
+
+        return x_0 * mean + noise * std
+
+    def q_mean_log_variance(self, x_0, t):
+        r"""Get mean and log_variance of diffusion process `q(x_t | x_0)`.
+
+        Args:
+            x_0 (torch.tensor): The original image before diffusion, shape as
+                [bz, ch, H, W].
+            t (torch.tensor): Target timestep, shape as [bz, ].
+
+        Returns:
+            Tuple(torch.tensor): Tuple contains mean and log variance.
+        """
+        tar_shape = x_0.shape
+        mean = var_to_tensor(self.sqrt_alphas_bar, t, tar_shape) * x_0
+        logvar = var_to_tensor(self.log_one_minus_alphas_bar, t, tar_shape)
+        return mean, logvar
+
+    def q_posterior_mean_variance(self, x_0, x_t, t, var=True, logvar=False):
+        r"""Get mean and variance of diffusion posterior
+            `q(x_{t-1} | x_t, x_0)`.
+
+        Args:
+            x_0 (torch.tensor): The original image before diffusion, shape as
+                [bz, ch, H, W].
+            t (torch.tensor): Target timestep, shape as [bz, ].
+            var (bool, optional): If set as ``True``, would return a dict
```

Comment:
This sentence does NOT contain the subject.

Reply:
`var` is not a suitable name for a bool variable. Please consider using `return_var`, `need_var`.